zookeeper-server/src/main/java/org/apache/zookeeper/server/backup/storage/impl/LocalBackupStorage.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.zookeeper.server.backup.storage.impl;
+
+import java.nio.file.Paths;
+
+import org.apache.zookeeper.common.ConfigException;
+import org.apache.zookeeper.server.backup.BackupConfig;
+
+public class LocalBackupStorage extends NfsBackupStorage {
```

Comment:
I have concerns about this class hierarchy. 

Is there a reason LocalBackupStorage is extending NfsBackupStorage? Conceptually, these two are different storage implementations - NFS is meant for NFS mounts, Local is meant for local server instances (say SSD or hard disk).

Reply:
There are no differences in the implementation for these two storages, just the paths where they point to, because the current implementation is general enough for all the file systems.
I agree that conceptually it is confusing to use this hierarchy. How about I refactor it a bit:
1. Abstract out a general file system backup storage, i.e. GeneralFsBackupStore
2. Both NfsBackupStorage and LocalBackupStorage will extend the GeneralFsBackupStore. So we don't need to duplicate code for these two storages. And since both NFS and local storage are file systems, I think it makes sense conceptually too.

Reply:
You have a few options:

1. As you said, you could have an abstract class (`PosixBackupStorage` for example) that implements BackupStorageProvider, and have both NFS and Local extend it. A potential issue with this is that HdfsBackupStorage is also a FS-based storage so it might be intrusive to an existing implementation. Another potential pitfall is that there could be different filesystem-based implementations, so you need to be careful about making sure that your abstract class's logic truly encompasses most, if not all, POSIX-based implementations.

2. This is the idea I shared with you previously. Just have a util class that provides static methods, which then could be used in both Local and NFS (Local and NFS directly implement the interface). This way, we 1) reduce duplicate code and 2) do not introduce unnecessary class hierarchy.

Reply:
Discussed offline. Since the current implementation is generic for different file systems, there is no need to have two different classes `NfsBackupStorage` and `LocalBackupStorage`. I would rename the `NfsBackupStorage` to `FileSystemBackupStorage`, and by using different paths, it will store files through NFS or to local storage.